apps/www/_blog/2025-05-08-event-triggers-wo-superuser.mdx

```diff
@@ -0,0 +1,119 @@
+
+## Privileged Role
+
+![utility hook](/images/blog/2025-05-08-event-triggers-wo-superuser/utility-hook.png)
+
+The core of `supautils` is the “privileged role”, which is a role that serves as proxy to superuser. It provides a safe subset of superuser capabilities and it’s accessible to end users.
+
+When the privileged role does a CREATE EVENT TRIGGER, we intercept the statement with an Utility Hook (`ProcessUtility_hook`). Here we elevate the role to a superuser, continuing the usual flow and allowing the creation on PostgreSQL core [1].
+
+Creating an event trigger like this is not safe though, as it would allow privilege escalation.
+
+## The privilege escalation problem
+
+Here a problem arises, once an event trigger is created:
+
+- It targets every role.
+- It runs using the target role privileges [2].
+
+This means that a malicious user can create an event trigger like:
+
+```sql
+create or replace function become_super()
+    returns event_trigger
+    language plpgsql as
+$$
+begin
+    alter role malicious superuser;
+end;
+$$;
+
+create event trigger malicious_event_trigger on ddl_command_end
+execute procedure become_super();
+```
+
+And once a SUPERUSER trips on the malicious event trigger, it will fire with its privileges. Making the malicious user a SUPERUSER.
+
+## Skipping Event Triggers
+
+![fmgr hook](/images/blog/2025-05-08-event-triggers-wo-superuser/fmgr-hook.png)
+
+A solution would be skipping user event triggers for superusers.
+
+The Function Manager hook (`fmgr_hook`) allows us to intercept and modify functions’ execution.
+
+We can intercept the event trigger function and replace it with a “noop”. PostgreSQL doesn’t provide a noop function, but we can use the existing `version()` function to achieve the same effect.
+
+Besides superusers, we also want to skip user event triggers for “reserved roles” [3]. These are used for managed services (like `pgbouncer`).
+
+## User Event Triggers
+
+This now allows users to safely create event triggers, without superuser access:
+
+```sql
+-- use the privileged role, which is configured to be "postgres"
+set role postgres;
+select current_setting('is_superuser'); -- prove it's not a superuser
+ current_setting
+-----------------
+ off
+(1 row)
+
+-- now create the event trigger
+create function show_current_user()
+returns event_trigger as $$
+begin
+  raise notice 'the event trigger is executed for %', current_user;
+end;
+$$ language plpgsql;
+
+create event trigger myevtrig on ddl_command_end
+execute procedure func();
+
+-- check it succeeds
+create table foo();
+NOTICE:  the event trigger is executed for privileged_role
+```
+
+## Future in PostgreSQL core
+
+We would also like to allow regular user event triggers in PostgreSQL core. To this end, we’ve submitted some patches which are already generating discussion.
+
+https://www.postgresql.org/message-id/flat/CAGRrpzbtYDkg7_xwfzrqByYgCJQbbL38tADyuN%2B6tAkbA-Pnkg%40mail.gmail.com
```

Comment:
[some patches]( 
https://www.postgresql.org/message-id/flat/CAGRrpzbtYDkg7_xwfzrqByYgCJQbbL38tADyuN%2B6tAkbA-Pnkg%40mail.gmail.com)   ?

and remove the link body?